ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java

```diff
@@ -216,60 +219,38 @@ public static void on_tick(Store.Transaction store, UnsignedLong time) {
     }
   }
 
-  public static boolean isFutureBlock(
-      BeaconBlock block, UnsignedLong genesisTime, UnsignedLong now) {
-    final UnsignedLong blockTime = getBlockTime(block, genesisTime);
-    return now.compareTo(blockTime) < 0;
-  }
-
-  private static UnsignedLong getBlockTime(BeaconBlock block, UnsignedLong genesisTime) {
-    return genesisTime.plus(block.getSlot().times(UnsignedLong.valueOf(SECONDS_PER_SLOT)));
-  }
-
   /**
    * @param store
    * @param block
    * @param st
    * @see
    *     <a>https://github.com/ethereum/eth2.0-specs/blob/v0.8.1/specs/core/0_fork-choice.md#on_block</a>
    */
-  public static BlockProcessingRecord on_block(
-      Store.Transaction store, BeaconBlock block, StateTransition st)
-      throws StateTransitionException {
-    // Make a copy of the state to avoid mutability issues
-    checkArgument(
-        store.containsBlockState(block.getParent_root()),
-        "on_block: Parent block state is not contained in block_state");
+  @CheckReturnValue
+  public static BlockImportResult on_block(
+      Store.Transaction store, BeaconBlock block, StateTransition st) {
     final BeaconState preState = store.getBlockState(block.getParent_root());
-    BeaconStateWithCache state =
-        BeaconStateWithCache.deepCopy(BeaconStateWithCache.fromBeaconState(preState));
 
-    // Blocks cannot be in the future. If they are, their consideration must be delayed until the
-    // are in the past.
-    checkArgument(
-        !isFutureBlock(block, state.getGenesis_time(), store.getTime()),
-        "on_block: Blocks cannot be in the future.");
+    // Return early if precondition checks fail;
+    final Optional<BlockImportResult> maybeFailure =
+        checkOnBlockPreconditions(block, preState, store);
+    if (maybeFailure.isPresent()) {
+      return maybeFailure.get();
+    }
 
     // Add new block to the store
     store.putBlock(block.signing_root("signature"), block);
 
-    checkArgument(
-        get_ancestor(
-                store,
-                block.signing_root("signature"),
-                store.getBlock(store.getFinalizedCheckpoint().getRoot()).getSlot())
-            .equals(store.getFinalizedCheckpoint().getRoot()),
-        "on_block: Check block is a descendant of the finalized block");
-
-    checkArgument(
-        block
-                .getSlot()
-                .compareTo(compute_start_slot_at_epoch(store.getFinalizedCheckpoint().getEpoch()))
-            > 0,
-        "on_block: Check that block is later than the finalized epoch slot");
+    // Make a copy of the state to avoid mutability issues
+    BeaconStateWithCache state =
+        BeaconStateWithCache.deepCopy(BeaconStateWithCache.fromBeaconState(preState));
```

Comment:
Another small change - we're now creating a copy of the state only after the block has been validated.